### PR TITLE
Add hermetic_cc_toolchain for a hermetic cc toolchain

### PR DESCRIPTION
**Pull request by prestonvanloon**
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

This PR introduces a replacement C/C++ toolchain for Prysm. Previously, we were using a LLVM toolchain from github.com/grailbio/bazel-toolchain and only when the user requested `--config=llvm`. 

This PR aims to solve two problems:

**Building C++ Depends on Host Tooling** -- When building anything that depends on C or C++ code in prysm, bazel will use the host machine gcc. This can cause undesired behavior where identical builds on different linux machines result in different outputs. Namely, if your host is using a recent version of gcc then non-backwards compatible glibc libraries will be requested. See https://github.com/prysmaticlabs/prysm/issues/7750 for an example. This behavior is non-hermetic and violates the desired principal of reproducible builds. 

**llvm Config Doesn't Work on Many OSes** -- The way that the existing cc toolchain works is that it downloads an appropriate version of llvm based on the host's OS. Some users have complained that the LLVM toolchain fails when using an operating system like Pop!OS or even Arch Linux. Again, this may be non-hermetic behavior as the host's OS should not influence how the CC toolchain compiles Prysm. 

The key features/changes are:
- [x] Default to using this cc_toolchain instead of the host / gcc toolchain
- [x] Pin glibc to 2.31 to support ubuntu 20.04 explicitly. (EOL is April, 2025)
- [x] Improve cross compilation by having a cc_toolchain that "just works". Enables support for multi-arch compilation for docker: https://github.com/prysmaticlabs/prysm/pull/12428

**Which issues(s) does this PR fix?**

Supercedes #12100 

Fixes https://github.com/prysmaticlabs/prysm/issues/7750

**Other notes for review**

**Testing plan**

To test this PR, we want to see that the release assets will build and everything else _should_ build, but some manual targets are expected to fail on some configurations. (Some things are linux only)
```
# Build release assets
bazel build --config=release //cmd/beacon-chain //cmd/validator //cmd/client-stats //cmd/prysmctl

# Everything should work
bazel build //...

# Optionally, build everything including manual tests
bazel build //... --build_manual_tests
```

- [x] Builds on linux amd64
- [ ] Builds on linux arm64 (untested)
- [ ] Builds on darwin amd64 (Mac OS without M1/M2 chip)  :x: Currently failing
- [ ] Builds on darwin arm64 (Mac OS with M1/M2 chip) :x: Currently failing

:x: Builds on windows amd64 (Not supported)

:warning: For darwin builds, I have added a hack that will allow a local C++ toolchain to be used until we can figure out how to hermetically provide a sysroot. We know this is possible with the docker cross compile, but we have to add that upstream to the toolchain provider. Additionally, there is a known issue where zig can't compile cgo on darwin. For the moment, we'll defer adding hermetic builds on darwin until those items can be resolved. 

Additionally, we want to see cross compile continues to work on from linux amd64
```
bazel build --config=release --config=linux_arm64_docker  //cmd/beacon-chain //cmd/validator //cmd/client-stats //cmd/prysmctl
bazel build --config=release --config=osx_amd64_docker  //cmd/beacon-chain //cmd/validator //cmd/client-stats //cmd/prysmctl
bazel build --config=release --config=osx_arm64_docker  //cmd/beacon-chain //cmd/validator //cmd/client-stats //cmd/prysmctl
bazel build --config=release --config=windows_amd64_docker  //cmd/beacon-chain //cmd/validator //cmd/client-stats //cmd/prysmctl
```

- [x] Cross compile still works on linux amd64

### CODE REVIEW

**Comment by prestonvanloon** [comment]
Waiting to merge this until after v4.0.5 next week.

File: .bazelrc [hunk 1]
```diff
@@ -27,30 +28,7 @@ build:minimal --@io_bazel_rules_go//go/config:tags=minimal
...
-build:llvm --incompatible_enable_cc_toolchain_resolution
-
-build:asan --copt -fsanitize=address,undefined
-build:asan --copt -fno-omit-frame-pointer
-build:asan --linkopt -fsanitize=address,undefined
-build:asan --copt -fno-sanitize=vptr,function
-build:asan --linkopt -fno-sanitize=vptr,function
-build:asan --copt -DADDRESS_SANITIZER=1
-build:asan --copt -D__SANITIZE_ADDRESS__
-build:asan --linkopt -ldl
```
Comment (kasey) [inline .bazelrc:44]: are we losing some build time safety checks, or are these somehow replicated in the hermetic_cc package?

File: build/bazelrc/zig-cc.bazelrc [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+# avoided. This option is not documented in bazel, so may break. If you intend
+# to use the hermetic toolchain exclusively, it won't hurt.
+build --action_env BAZEL_DO_NOT_DETECT_CPP_TOOLCHAIN=1
+
+# This snippet instructs Bazel to use the registered "new kinds of toolchains".
+# This flag not needed after this issue is closed https://github.com/bazelbuild/bazel/issues/7260
+build --incompatible_enable_cc_toolchain_resolution
+
+# Add a no-op warning for users still using --config=llvm
+build:llvm --unconditional_warning="llvm config is no longer used as clang is now the default compiler"
```
Comment (james-prysm) [inline build/bazelrc/zig-cc.bazelrc:19]: we don't need to mention zig here at all right? even though it's using clang underneath?

File: WORKSPACE [hunk 3]
```diff
@@ -16,28 +16,31 @@ load("@rules_pkg//:deps.bzl", "rules_pkg_dependencies")
...
-    name = "llvm_toolchain",
-    llvm_version = "13.0.1",
+# Register zig sdk toolchains with support for Ubuntu 20.04.
+# For ubuntu glibc support, see https://launchpad.net/ubuntu/+source/glibc
+register_toolchains(
+    "@zig_sdk//toolchain:linux_amd64_gnu.2.31",
+    "@zig_sdk//toolchain:linux_arm64_gnu.2.31",
+    "@zig_sdk//toolchain:darwin_amd64",
+    "@zig_sdk//toolchain:darwin_arm64",
+    "@zig_sdk//toolchain:windows_amd64",
```
Comment (kasey) [inline WORKSPACE:41]: +1 to your suggestion in huddle to rename this to match the upstream repo.

File: build/bazelrc/zig-cc.bazelrc [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# From Bazel's perspective, this is almost equivalent to always specifying
```
Comment (kasey) [inline build/bazelrc/zig-cc.bazelrc:1]: another thing to rename `zig-cc.bazelrc`

File: WORKSPACE (hunk 3 above)
Comment (prestonvanloon) [reply WORKSPACE:41]: I can't rename this one, it's a static value here: https://github.com/uber/hermetic_cc_toolchain/blob/bbd965d3ab0803475bd3bb9b59fb50c3451b5c71/toolchain/defs.bzl#L79

File: build/bazelrc/zig-cc.bazelrc (hunk 2 above)
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply build/bazelrc/zig-cc.bazelrc]: It is just a warning. Without some `build:llvm`, the build will fail entirely when using `--config=llvm`

**Comment by prestonvanloon** [comment]
darwin arm64 currently failing with 
```
external/go_sdk/pkg/tool/darwin_arm64/link: running external/zig_sdk/tools/aarch64-macos-none/c++ failed: exit status 1
warning(link): framework not found for '-framework CoreFoundation'
warning(link): Framework search paths:
error: FrameworkNotFound
```

Looks like https://github.com/ziglang/zig/issues/10513

**Comment by prestonvanloon** [comment]
Another blocker for mac amd64 https://github.com/ziglang/zig/issues/15438

**Review by james-prysm — APPROVED** [review]
was building fine locally so approving this

**Review by kasey — APPROVED** [review]

**Pull request merged**